dev (#1)
dovgopoly/u512-md: README.md

# The most expensive Solidity function

> Have you ever run into a function that just won’t fit into an Ethereum block because it costs too much gas? This is the story of how we tackled that very challenge!

Reviewer: Probably "... run into a situation where a smart contract function just won't ..." not "run into function".

Costs -> consumes. Gas cost has nothing to do with block size.

## The problem

> It all started when we built the Rarimo protocol. If you aren’t familiar with Rarimo, check out [this post](https://medium.com/@denys.riabtsev/national-passports-verification-with-zkp-part-1-b653f5e5c8d8) for more details. Rarimo lets users prove that they have a passport from a certain country without revealing any personal information.

Reviewer: Technically Rarimo protocol is not complete. Better rephrase "It all started during the development of the Rarimo protocol."

> It all started when we built the Rarimo protocol. If you aren’t familiar with Rarimo, check out [this post](https://medium.com/@denys.riabtsev/national-passports-verification-with-zkp-part-1-b653f5e5c8d8) for more details. Rarimo lets users prove that they have a passport from a certain country without revealing any personal information.

Reviewer: ```suggestion
It all started when we built the Rarimo protocol. If you aren’t familiar with Rarimo, check out [this post](https://medium.com/@denys.riabtsev/national-passports-verification-with-zkp-part-1-b653f5e5c8d8) for more details. Rarimo lets users prove via ZKP that they have a passport from a certain country without revealing any personal information. 
```

IDK, would be great to add one more sentence to indicate Rarimo's significance.

> The challenge was that there are no strict rules for passport cryptography. Each country can use its own method. Some use RSA or ECDSA signatures combined with various hash functions like SHA-256, SHA-384, SHA-512, and more. OpenPassport even has a [map](https://map.openpassport.app/) showing the algorithms used in passports worldwide. For example, Peru uses only ECDSA signatures over 384-bit curves, while Brazil uses 512-bit curves.

Reviewer: The challenge was that we had to support all this zoo, not just that these algos exist.

> At first, it seemed simple to build a solution that supports all these algorithms since the only real difference was the size of the parameters used. But here’s the problem: the Ethereum Virtual Machine (EVM) only supports 256-bit unsigned integers! So, we had three choices:

Reviewer: The readers will be educated enough

```suggestion
At first, it seemed simple to build a solution that supports all these algorithms since the only real difference was the size of the parameters used. But here’s the problem: the EVM only supports 256-bit unsigned integers! So, we had three choices:
```

1. **Use Zero-Knowledge Proofs:** Verify the signature with ZK-proofs. The downside was that users would need to download a huge "zkey", which would make the UX worse.
2. **Make a Precompile:** Create a precompile for signature verification. However, this would require a chain fork, which is not easy.

> 3. **Do Everything On-Chain:** Implement all the logic on-chain. Although this method wouldn’t work on Mainnet because of high gas costs, it was a good fit for us since we use our own Layer 2 solution with cheap gas. Our only challenge was to keep the function’s gas usage below 30 million per block.

Reviewer: ```suggestion
3. **Do Everything On-Chain:** Implement all the logic on-chain. Although this method wouldn’t work on Mainnet because of high gas costs, it was a good fit for us since Rarimo is a Layer 1 where gas is cheap. Also, the full-on-chain library could be easily reusable if required. Our only challenge was to keep the function’s gas usage below 30 million per block.
```

## BigNumber library

> Finally, we decided to go with the third approach: implementing ECDSA for both 384-bit and 512-bit curves directly on-chain. But we ran into another problem. It turned out that there isn’t an efficient "bigint" library available. Not because no one can write one, but because it’s nearly impossible to build one in a general way that works with any integer size while keeping the gas cost as low as possible.

Reviewer: ```suggestion
The choice was obvious, so we decided to go with the third approach: implementing ECDSA for both 384-bit and 512-bit curves directly on-chain. But we ran into another problem. It turned out that there isn’t an efficient "bigint" library available. Not because no one can write one, but because it’s nearly impossible to build one in a general way that works with any integer size while keeping the gas cost as low as possible.
```

> Finally, we decided to go with the third approach: implementing ECDSA for both 384-bit and 512-bit curves directly on-chain. But we ran into another problem. It turned out that there isn’t an efficient "bigint" library available. Not because no one can write one, but because it’s nearly impossible to build one in a general way that works with any integer size while keeping the gas cost as low as possible.

Reviewer: There is a problem with tenses here.

```suggestion
Finally, we decided to go with the third approach: implementing ECDSA verification for both 384-bit and 512-bit curves directly on-chain. But we ran into another problem. It turned out that there wasn't an efficient "bigint" library available. Not because no one could write one, but because it’s nearly impossible to implement one in a general way that works with arbitrary integer size while keeping the gas cost as low as possible.
```

> For example, we tried to integrate our code with the [BigNumber library](https://github.com/firoorg/solidity-BigNumber). This library is really cool and gave us ideas for future improvements, but our verification algorithm still used too much gas because of two issues:

Reviewer: ```suggestion
For example, we tried to integrate our Solidity code with the [BigNumber library](https://github.com/firoorg/solidity-BigNumber). This library is really cool and gave us ideas for future improvements, but our verification algorithm still used too much gas because of two issues:
```

1. **Abstraction costs:** The library is written in a general way to handle any integer size. However, as is often the case, any abstractions make your code less optimized.

> 2. **Memory Expansion:** The EVM doesn't have a heap or allocators for efficient memory management. Once memory is allocated, it is never released, which also results in [memory expansion](https://www.evm.codes/about#memoryexpansion) overhead.

Reviewer: I would say this is Solidity problem, not EVM.

> We spent a lot of time looking for a solution. We tried to optimize the library and even emulated a heap and allocators directly in Solidity to cut down the overhead. We even found some bugs and  [contributed](https://github.com/firoorg/solidity-BigNumber/pull/18)  to the original BigNumber library.

Reviewer: ```suggestion
We spent a lot of time looking for a solution. We tried to optimize the library and even emulated a heap and allocators directly in Solidity to cut down the overhead. We even found some bugs and [contributed](https://github.com/firoorg/solidity-BigNumber/pull/18) to the original BigNumber library.
```

> But nothing worked. Our implementation still required more than 400 billion gas for execution plus gas needed for memory expansion overhead. We couldn’t even run the whole algorithm in our testing environment because the machine ran out of memory.

Reviewer: ```suggestion
But nothing worked. Our implementation still required more than 400 billion gas for execution plus gas needed for memory expansion overhead. We couldn’t even run the whole algorithm in our testing environment because the machine ran out of RAM!
```

## Uint512

Just when we were about to give up and try one of the other options, a sudden idea struck us. We realized that we didn't need a general BigNumber library at all - we only needed 512-bit integers! With everything we had learned from working with the BigNumber library, we believed it's possible to implement our own highly optimized library for 512-bit unsigned integers.

### Memory layout

> How to implement a uint512 library most efficiently? The first challenge was deciding how to store a uint512 in memory. Since the integer size is fixed, we can always use 2 memory slots - each 256 bits - to hold a uint512. But what data type should be used? Structs? Bytes? Neither, because both require an extra memory slot. For example, a bytes type needs a slot for its length, and a struct needs a slot for a data pointer.

Reviewer: Structs require a data pointer only if they store a dynamic element. Otherwise they are inlined. I would rephrase that Solidity compiler creates a problem because all these "memory" variables are always reallocated

That's why we decided to use raw pointers directly on the stack. They are automatically removed when the function finishes, so only 2 memory slots are used for each uint512. Using less memory reduces the overhead from memory expansion, which is crucial for functions that perform a lot of work.

The diagram below shows the different storage structures:

![diagram](./layout.png)

> Even more, starting with Solidity 0.8.8, [user-defined value types](https://soliditylang.org/blog/2021/09/27/user-defined-value-types/) are added, which makes the user experience much smoother. You no longer need to use the `memory` keyword, and you might even forget that the uint512 type isn’t built-in!

Reviewer: ```suggestion
Even more, starting with Solidity 0.8.8, [user-defined value types](https://soliditylang.org/blog/2021/09/27/user-defined-value-types/) were added, which makes the user experience much smoother. You no longer need to use the `memory` keyword, and you might even forget that the uint512 type isn’t built-in!
```

Here's an example:

```
// U512.sol
type uint512 is uint256; // synonym for pointer

// Contract.sol
uint512 a_ = U512.fromUint256(3);
uint512 b_ = U512.fromUint256(6);
uint512 m_ = U512.fromUint256(5);
uint512 r_ = U512.modadd(a_, b_, m_);
U512.eq(r_, U512.fromUint256(4)); // true
U512.toBytes(r_); // "0x00..04" - 64 bytes

> ```

Reviewer: I'd make a separate section "result" and showcase it there.

For example, when adding two uint512 values, you add the lower words first and then the upper words, making sure to handle any overflow from the lower word addition. The diagram below shows how you can implement the add function.

![diagram](./add.png)

### Modexp magic 

But what about mod operations? Is it necessary to write a mod function in Solidity? That would be tough. The answer is no. We can use the [modexp precompile](https://www.evm.codes/precompiled?fork=cancun#0x05) available in the EVM! This built-in feature lets you make a static call to a predefined contract that calculates `b^e mod m` for any size of input.

This saves a lot of gas because, instead of coding the mod algorithm on-chain, the Ethereum node does the heavy work for you at a low cost. With this tool, it's easy to implement functions like modexp directly, as well as mod (by setting `e = 1`), and modadd along with modmul by first computing the result and then taking it modulo.

> However, you must use the modexp precompile carefully. For example, we once accidentally passed the exponent size as 64 bytes, which ended up costing 10 times more gas, even though the exponent was just 1!

Reviewer: I'd rephrase like:

However, you must use the modexp precompile carefully. For example, passing the exponent size as 64 bytes would result in the precompile costing 10 times more gas, even if the exponent was small!

Fortunately, this is a known problem that can be solved efficiently using the EVM’s [modmul opcode](https://www.evm.codes/?fork=cancun#09). Remco Bloemen explains this approach in his article [Mathemagic finale: muldiv](https://xn--2-umb.com/21/muldiv/). We don’t need the division part of his method, but the small piece of code that calculates the most significant 256 bits of the product - "huge carry". It's only three lines of code, but it packs a lot of math:

```
// 512-bit multiply [prod1 prod0] = a * b
// Compute the product mod 2**256 and mod 2**256 - 1
// then use the Chinese Remainder Theorem to reconstruct
// the 512 bit result. The result is stored in two 256
// variables such that product = prod1 * 2**256 + prod0
uint256 prod0; // Least significant 256 bits of the product
uint256 prod1; // Most significant 256 bits of the product
assembly {
	let mm := mulmod(a, b, not(0))
	prod0 := mul(a, b)
	prod1 := sub(sub(mm, prod0), lt(mm, prod0))
}
```

> This approach brings the gas cost of the multiplication function very close to that of the addition function: 353 gas (mul) vs 256 gas (add). Similarly, since only one modexp call is needed, the modulo versions become much more efficient 1176 gas (modmul) 760 gas (modadd).

Reviewer: That's great, but let's add a sentence that this method has to be "translated" to 512-bit numbers

```  
call512 call_ = U512.initCall();  
uint512 a_ = U512.fromUint256(3);  
uint512 b_ = U512.fromUint256(6);  
uint512 m_ = U512.fromUint256(5);  
uint512 r_ = U512.modadd(call_, a_, b_, m_); // 4  
U512.modmulAssign(call_, r_, a_, m_); // assigns (r_ * a_) % m_ to r_
r_.eq(U512.fromUint256(2)); // true
r_.toBytes(); // "0x00..02"  
```

## Gas costs

By applying all the optimizations, we achieved some impressive results:

| uint512 operation | Avg gas  |
| ----------------- | -------- |
| add               | 269 gas  |
| sub               | 278 gas  |
| mul               | 353 gas  |
| mod               | 682 gas  |
| modinv            | 6083 gas |
| modadd            | 780 gas  |
| redadd            | 590 gas  |
| modmul            | 1176 gas |
| modsub            | 1017 gas |
| redsub            | 533 gas  |
| modexp            | 5981 gas |
| modexpU256        | 692 gas  |
| moddiv            | 7092 gas |
| and               | 251 gas  |
| or                | 251 gas  |
| xor               | 251 gas  |
| not               | 216 gas  |
| shl               | 272 gas  |
| shr               | 272 gas  |

> For the ECDSA algorithm, the 384-bit curves consume 8.9 million gas and the 512-bit curves consume 13.6 million gas - a **30,000x improvement** compared to the initial version!

Reviewer: Need to add a section

## Implementation

You can find the implementation on Github, blah, blah, link
